oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/web/authentication/logout/OidcBackchannelLogoutFilter.java

```diff
@@ -0,0 +1,206 @@
+import org.springframework.security.oauth2.jwt.JwtDecoder;
+import org.springframework.security.oauth2.jwt.JwtDecoderFactory;
+import org.springframework.security.web.authentication.logout.BackchannelLogoutAuthentication;
+import org.springframework.security.web.authentication.logout.BackchannelLogoutHandler;
+import org.springframework.security.web.authentication.logout.LogoutHandler;
+import org.springframework.security.web.util.matcher.AntPathRequestMatcher;
+import org.springframework.security.web.util.matcher.RequestMatcher;
+import org.springframework.util.Assert;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+/**
+ * A filter for the Client-side OIDC Backchannel Logout endpoint
+ *
+ * @author Josh Cummings
+ * @since 6.1
+ * @see <a target="_blank" href=
+ * "https://openid.net/specs/openid-connect-backchannel-1_0.html">OIDC Backchannel Logout
+ * Spec</a>
+ */
+public class OidcBackchannelLogoutFilter extends OncePerRequestFilter {
+
+	private final Log logger = LogFactory.getLog(getClass());
+
+	private static final String ERROR_MESSAGE = "{ \"error\" : \"%s\", \"error_description\" : \"%s\" }";
+
+	private final ClientRegistrationRepository clientRegistrationRepository;
+
+	private final JwtDecoderFactory<ClientRegistration> logoutTokenDecoderFactory;
+
+	private RequestMatcher requestMatcher = new AntPathRequestMatcher("/oauth2/{registrationId}/logout", "POST");
+
+	private OidcProviderSessionRegistry providerSessionRegistry = new InMemoryOidcProviderSessionRegistry();
+
+	private LogoutHandler logoutHandler = new BackchannelLogoutHandler();
+
+	/**
+	 * Construct an {@link OidcBackchannelLogoutFilter}
+	 * @param clients the {@link ClientRegistrationRepository} for deriving Logout Token
+	 * validation
+	 * @param logoutTokenDecoderFactory the {@link JwtDecoderFactory} for constructing
+	 * Logout Token validators
+	 */
+	public OidcBackchannelLogoutFilter(ClientRegistrationRepository clients,
+			JwtDecoderFactory<ClientRegistration> logoutTokenDecoderFactory) {
+		this.clientRegistrationRepository = clients;
+		this.logoutTokenDecoderFactory = logoutTokenDecoderFactory;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+			throws ServletException, IOException {
+		RequestMatcher.MatchResult result = this.requestMatcher.matcher(request);
+		if (!result.isMatch()) {
+			chain.doFilter(request, response);
+			return;
+		}
+		String logoutToken = request.getParameter("logout_token");
+		if (logoutToken == null) {
+			this.logger.debug("Did not process OIDC Backchannel Logout since no logout token was found");
+			chain.doFilter(request, response);
+			return;
+		}
+		String registrationId = result.getVariables().get("registrationId");
+		ClientRegistration registration = this.clientRegistrationRepository.findByRegistrationId(registrationId);
+		if (registration == null) {
```

Comment:
> Can you please clarify specifically which aspects you are wanting to see moved down a layer?

Take a look at `OAuth2LoginAuthenticationFilter` and you'll see it instantiates an `OAuth2LoginAuthenticationToken`, which includes all the request data and then it's passed to `OAuth2LoginAuthenticationProvider` OR `OidcAuthorizationCodeAuthenticationProvider` indirectly via `AuthenticationManager`. The `AuthenticationProvider` is responsible for validating the request data to ensure it's valid and then returns an "authenticated" `Authentication` result object. It's the standard pattern we use for authenticating credentials. The `Filter` acts as a simple `Controller` which performs minimal data validation/ sanitation but the `AuthenticationProvider` performs the "business rule" validation. For example, I would expect the `logout_token` parameter to be validated in the `AuthenticationProvider` using the `JwtDecoder` and NOT in the `Filter`.

Reply:
Sorry, I'm still not quite clear. I understand the model as we use it in many places in the codebase. What I'm not clear on is what logic you are specifically proposing be moved down one layer.

It seems like you are asking that this:

```java
OidcLogoutToken token;
try {
	token = token(logoutToken, registration);
}
catch (BadJwtException ex) {
	this.logger.debug("Failed to process OIDC Backchannel Logout", ex);
	String message = String.format(ERROR_MESSAGE, "invalid_request", ex.getMessage());
	response.sendError(400, message);
	return;
}
```

change to this:

```java
OidcLogoutToken token;
try {
	token = authenticate(logoutToken, registration);
}
catch (AuthenticationException ex) {
	this.logger.debug("Failed to process OIDC Backchannel Logout", ex);
	String message = String.format(ERROR_MESSAGE, "invalid_request", ex.getMessage());
	response.sendError(400, message);
	return;
}
```

where `authenticate` points to an `AuthenticationProvider`.

If I've understood you correctly, then why add a layer to abstract away a single component? My concern is that with an additional layer, configuring the `JwtDecoderFactory` would require configuring two components (the `JwtDecoderFactory` and the `AuthenticationProvider`). Is there a known use case where this extra configuration burden is the correct tradeoff? If not, I'd vote to keep the object hierarchy as simple as possible and then abstract further down the road when the need is apparent.

Reply:
Actually, I think an important key to consider is your comment on another thread:

> I believe the reason why we didn't go with `IdTokenClaimAccessor extends JwtClaimAccessor` is because we would introduce a required dependency of `oauth2-client` -> `oauth2-jose`

As such, I'll go ahead and abstract it away. Given the reasons I already stated, I'll update the PR to use `AuthenticationManager` instead of `AuthenticationProvider` and we can look at that together. While I'm not moved by arguments of consistency with other OAuth filters, if there is a compelling reason to use `AuthenticationProvider`, I don't mind changing.